src/java/main/org/apache/zookeeper/server/quorum/CommitProcessor.java

```diff
@@ -251,27 +252,29 @@ public void run() {
                     LinkedList<Request> sessionQueue = pendingRequests
                             .get(request.sessionId);
                     if (sessionQueue != null) {
+                        Request topPending = sessionQueue.peekFirst();
                         // If session queue != null, then it is also not empty.
-                        Request topPending = sessionQueue.poll();
-                        if (request.cxid != topPending.cxid) {
-                            LOG.error(
-                                    "Got cxid 0x"
-                                            + Long.toHexString(request.cxid)
-                                            + " expected 0x" + Long.toHexString(
-                                                    topPending.cxid)
-                                    + " for client session id "
-                                    + Long.toHexString(request.sessionId));
-                            throw new IOException("Error: unexpected cxid for"
-                                    + "client session");
+                        // we can get commit requests for requests before we even get this session
+                        //(see ZOOKEEPER-2684). We will just pass those commits to the next processor.
```

Comment:
please remove the new comments from here and add them inside the if, where this case is handled.